src/wp-includes/class-wp-block-scanner.php

```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @var int
+	 */
+	private $delimiter_length = 0;
+
+	/**
+	 * Byte offset into source text after which last matched block ended.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $last_delimiter_ended_before = 0;
+
+	/**
+	 * Byte offset where namespace span begins.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $namespace_at = 0;
+
+	/**
+	 * Byte length of namespace span, or `0` if implicitly in the "core" namespace.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $namespace_length = 0;
+
+	/**
+	 * Byte offset where block name span begins.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $name_at = 0;
+
+	/**
+	 * Byte length of block name span.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $name_length = 0;
+
+	/**
+	 * Whether the delimiter contains the block-closing flag.
+	 *
+	 * This may be erroneous if present within a void block,
+	 * therefore the {@see self::has_closing_flag} can be used by
+	 * calling code to perform appropriate error-handling.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var bool
+	 */
+	private $has_closing_flag = false;
+
+	/**
+	 * Byte offset where JSON attributes span begins.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $json_at;
+
+	/**
+	 * Byte length of JSON attributes span, or `0` if none are present.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var int
+	 */
+	private $json_length = 0;
+
+	/**
+	 * Internal parser state, differentiating whether the instance is currently matched,
+	 * on an implicit freeform node, in error, or ready to begin parsing.
+	 *
+	 * @since {WP_VERSION}}
+	 *
+	 * @var string
+	 */
+	private $state = self::READY;
+
+	/**
+	 * Indicates what kind of block comment delimiter this represents.
+	 *
+	 * One of:
+	 *
+	 *  - `static::OPENER` If the delimiter is opening a block.
+	 *  - `static::CLOSER` If the delimiter is closing an open block.
+	 *  - `static::VOID`   If the delimiter represents a void block with no inner content.
+	 *
+	 * If a parsed comment delimiter contains both the closing and the void
+	 * flags then it will be interpreted as a void block to match the behavior
+	 * of the official block parser, however, this is a mistake and probably
+	 * the block ought to close an open block of the same name, if one is open.
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @var string
+	 */
+	private $type;
+
+	/**
+	 * Creates a new block scanner.
+	 *
+	 * Example:
+	 *
+	 *     $scanner = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter() ) {
+	*          if ( $scanner->opens_block( 'core/image' ) ) {
+	 *             echo "Found an image!\n";
+	 *         }
+	 *     }
+	 *
+	 * @see self::next_delimiter
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( $source_text ) {
```

Comment:
```suggestion
	public static function create( $source_text ): ?self {
```

Reply:
Although, if the `string` type hint is added to `$source_text` then a fatal error would have occurred, and this method would never return anything but `self`:

```suggestion
	public static function create( string $source_text ): self {
```

Reply:
leaving this open with `?self` so that we can fail it in the future if we decide to. otherwise this and every subclass will be prevented from adding conditions to creating the scanner.

Reply:
this method is gone, so this comment is no longer relevant.